src/components/CalendarContainer.ts

```diff
@@ -0,0 +1,293 @@
+import { Component, createElement } from "react";
+
+import { Alert } from "./Alert";
+import { Calendar, CalendarEvent, View } from "./Calendar";
+
+interface WrapperProps {
+    mxform: mxui.lib.form._FormBase;
+    mxObject: mendix.lib.MxObject;
+}
+
+export interface CalendarContainerProps extends WrapperProps {
+    titleAttribute: string;
+    startAttribute: string;
+    endAttribute: string;
+    defaultView: View;
+    dataSource: DataSource;
+    eventEntity: string;
+    entityConstraint: string;
+    dataSourceMicroflow: string;
+    popup: boolean;
+    selectable: boolean;
+    onClickEvent: OnClickEventOptions;
+    onClickSlotEvent: OnClickEventOptions;
+    eventPage: string;
+    openPageAs: PageLocation;
+    eventMicroflow: string;
+    eventNanoflow: Nanoflow;
+    slotPage: string;
+    slotMicroflow: string;
+    slotNanoflow: Nanoflow;
+}
+
+type DataSource = "XPath" | "microflow";
+type OnClickEventOptions = "doNothing" | "showPage" | "callMicroflow" | "callNanoflow";
+type PageLocation = "content" | "popup" | "modal";
+
+interface CalendarContainerState {
+    alertMessage?: string;
+    events: CalendarEvent[];
+}
+
+interface Nanoflow {
+    nanoflow: object[];
+    paramsSpec: { Progress: string };
+}
+
+export default class CalendarContainer extends Component<CalendarContainerProps, CalendarContainerState> {
+    private subscriptionHandles: number[];
+
+    constructor(props: CalendarContainerProps) {
+        super(props);
+
+        this.subscriptionHandles = [];
+        this.state = {
+            alertMessage: CalendarContainer.validateProps(this.props),
+            events: []
+        };
+    }
+
+    render() {
+        if (this.state.alertMessage) {
+            return createElement(Alert, {
+                bootstrapStyle: "danger",
+                className: "widget-calendar-alert",
+                message: this.state.alertMessage
+            });
+        } else {
+            return createElement(Calendar, {
+                events: this.state.events,
+                defaultView: this.props.defaultView,
+                popup: this.props.popup,
+                selectable: this.props.selectable,
+                onSelectEventAction: this.onClickEvent,
+                onSelectSlotAction: this.onClickSlot,
+                onEventDropAction: this.onDropEvent
+            });
+        }
+    }
+
+    componentDidMount() {
+        if (!this.state.alertMessage && this.props.mxObject) {
+            this.fetchData(this.props.mxObject);
+        }
+    }
+
+    componentWillUnMount() {
+        this.subscriptionHandles.forEach(window.mx.data.unsubscribe);
+    }
+
+    componentWillReceiveProps(nextProps: CalendarContainerProps) {
+        this.fetchData(nextProps.mxObject);
+        this.resetSubscriptions(nextProps.mxObject);
+    }
+
+    private fetchData = (mxObject: mendix.lib.MxObject) => {
+        const { dataSource, dataSourceMicroflow, eventEntity } = this.props;
```

Comment:
Not adding much value. Inline `this.props`

Reply:
Done